### ISSUE TITLE
Upload watchdog events

### ISSUE THREAD

**Issue by KevinWassermann94** [NebraLtd/hm-diag#395]
@KevinWassermann94 commented on [Wed Jun 22 2022](https://github.com/NebraLtd/hm-watchdog/issues/6)

Mechanism for copying data into BigQuery should be able to be extended to also copy the data to the Dashboard. Some options: Cloud Storage -> Functions -> PubSub -> BigQuery/Dashboard, OR Storage -> Functions -> BigQuery -> Dashboard, OR Storage -> Function -> Dashboard and separately to BigQuery.

**Acceptance Criteria:**
- [ ] Payload is JSON containing system uptime(at time of the event), local connectivity, internet connectivity, watchdog event(network manager restart, device reboot, forced reboot), duration of failures(on network manager restart, reboot), timestamp, serial number, variant, firmware version, region override, uptime of containers, container status, packet loss for network interfaces, uuid
- [ ] Uploads can be in a public location 
- [ ] It should be possible to create analytics from this data
- [ ] Upload events as soon as possible, store logs offline when miner doesn't have connectivity
- [ ] Upload on every network watchdog runtime(every hour)
- [ ] Create analytics of the average uptime by day
- [ ] Upload not at the flat hour, rather random during the hour

**Comment by pritamghanghas** [comment]
[network_watchdog_events.csv](https://github.com/NebraLtd/hm-diag/files/9179926/network_watchdog_events.csv)

**Comment by pritamghanghas** [comment]
@marvinmarnold I was using my personal gcloud account for development. This now needs a bucket named "helium-miner-events" with public write. 
Requires corresponding [function](https://github.com/NebraLtd/hm-diag/blob/master/bigquery/events_import_script.py) as well as big query table with [schema](https://github.com/NebraLtd/hm-diag/blob/master/bigquery/events_schema.json)

**Comment by htsili** [comment]
@pritamghanghas what should be the name of the big query table?

**Comment by pritamghanghas** [comment]
@htsili We can keep it like diagnostics. we will have to update the function code file accordingly. But that is also similar to already running function for diagnostics.
<dataset>.<table>
hotspot_events_data.events

**Comment by htsili** [comment]
@pritamghanghas 
do we need to create a new dataset called hotspot_events_data or are we going to put it under hotspot_diagnostics_data?

**Comment by htsili** [comment]
@pritamghanghas i have created the table called events under the hotspot_events_data dataset.

**Comment by pritamghanghas** [comment]
@htsili this needs corresponding bucket and triggered function. 
This is how the data is supposed to flow. It is same in case of diag for reference.
Hotspot uploads to public bucket -> upload triggers function -> functions write to bigquery.

**Comment by pritamghanghas** [comment]
@htsili 
while copying the function [these lines](https://github.com/NebraLtd/hm-diag/blob/master/bigquery/events_import_script.py#L10) will have to change according to the big query table names. Ping me when you do it.

**Comment by marvinmarnold** [comment]
@pritamghanghas can you make sure you initialize logging, like this: https://github.com/NebraLtd/product-management/blob/20911395b377a86813e7b18e56484ab250ff22e9/analysis/chargify-read-site-stats/main.py#L17

**Comment by kashifpk** [comment]
@marvinmarnold @KevinWassermann94 @pritamghanghas - once this is done, let's create a follow-up ticket so we can also stream this data through our broker to the dashboard.

**Comment by KevinWassermann94** [comment]
@kashifpk This will be the follow up: https://github.com/NebraLtd/hm-dashboard/issues/1338

**Comment by pritamghanghas** [comment]
@KevinWassermann94 @marvinmarnold I was trying to get the 1 hour heartbeat done. But got doubts. These events were supposed to be stored in case of failure to upload due to network or otherwise. How do we plan to use them for uptime if they always arrive. 

currently merged code: network state changes are published and network is checked every hour.

**Comment by htsili** [comment]
@pritamghanghas I have created the bucket helium-miner-events.
What should be the name of the cloud function?

**Comment by pritamghanghas** [comment]
cloud function name can be anything, it should be triggered on object creation in the bucket just like diagnostics. eg. injest_hotspot_events

**Comment by htsili** [comment]
@pritamghanghas i have created the function [helium-miner-injest-hotspot-events](https://console.cloud.google.com/functions/details/us-central1/helium-miner-injest-hotspot-events?env=gen1&authuser=0&orgonly=true&project=nebra-production)

**Comment by pritamghanghas** [comment]
@htsili I can't do any edits on it. Wanted to update the schema with [latest](https://github.com/NebraLtd/hm-diag/blob/pritam/watchdog_heartbeat/bigquery/events_schema.json). It has this extra field now.

  {
    "name": "network_state",
    "type": "STRING",
    "mode": "NULLABLE"
  }

**Comment by htsili** [comment]
@pritamghanghas i have added it for you.

**Comment by pritamghanghas** [comment]
@htsili the bucket has to public writable.

```
DEBUG:urllib3.connectionpool:https://www.googleapis.com:443 "POST /upload/storage/v1/b/helium-miner-events/o?uploadType=media&name=-1659069341.714651.json HTTP/1.1" 401 438
ERROR:root:Error submitting event to GCS bucket: 401 - {
  "error": {
    "code": 401,
    "message": "Anonymous caller does not have storage.objects.create access to the Google Cloud Storage object.",
    "errors": [
      {
        "message": "Anonymous caller does not have storage.objects.create access to the Google Cloud Storage object.",
        "domain": "global",
        "reason": "required",
        "locationType": "header",
        "location": "Authorization"
      }
    ]
  }
}
```

**Comment by htsili** [comment]
@pritamghanghas why does the bucket need to be public writable? That seems a bit risky.

**Comment by pritamghanghas** [comment]
Brief implementation details and tail of logs verifying that are recorded [here](https://docs.google.com/document/d/1tel4waTol9uao-Zn5fPAs7l6KYm2kgahOf8_C3R82ng/edit?usp=sharing)

**Comment by marvinmarnold** [comment]
@pritamghanghas were heartbeats implemented ultimately? Will we continue to get events even if the miner remains online?

**Comment by marvinmarnold** [comment]
Here are two queries that seem to provide useful info based on this sample data

```
SELECT
  day,
  avg(max_uptime) as avg_uptime,
  max(max_uptime) as max_uptime
FROM (
  SELECT
    day,
    serial,
    max(uptime_hours) as max_uptime
  FROM (
    SELECT 
      date(generated_ts) as day,
      serial,
      uptime_hours
    FROM `nebra-production.hotspot_events_data.events`
    where event_type = 'HEARTBEAT'
  )
  group by serial, day
)
group by day
```
![Screen Shot 2022-08-02 at 2 14 51 PM](https://user-images.githubusercontent.com/1311030/182455156-a78a3caa-3e2a-4487-826d-cef4ac4713a0.png)



```
-- Num state changes by day
SELECT
  event_type,
  day,
  count(*) as num_events,
  avg(uptime_hours) as avg_uptime
FROM (
  SELECT 
    date(generated_ts) as day,
    event_type,
    uptime_hours
  FROM `nebra-production.hotspot_events_data.events`
  where event_type != 'HEARTBEAT'
)
group by event_type, day
```

![Screen Shot 2022-08-02 at 2 11 54 PM](https://user-images.githubusercontent.com/1311030/182454679-adb26dc9-1279-4776-9a70-7df73983d64d.png)

**Comment by KevinWassermann94** [comment]
The region_override seems to be empty on all logs.
I'm currently verifying the failed containers state.

Could you clarify what packet_errors mean? E.g. if packet_errors = 446 what does that number imply?
It seems like it is retained over runtime?

**Comment by pritamghanghas** [comment]
I check helium_testnet fleet doesn't use region_override. It will be other than empty only if it is set. 
packet_error is total wifi/ethernet errors reported by kernel. Yes this number will keep accumulating till a reboot.

**Comment by KevinWassermann94** [comment]
The devices have a local region_override they receive from the Helium miner. It's the frequency they operate in

**Comment by KevinWassermann94** [comment]
Failed containers are showing properly. I am happy to merge and fix the region_override seperately
